### PR TITLE
WT-2307: diagnose split next bug

### PR DESCRIPTION
**Pull request by agorrod**
Add in diagnostic code to track cursor operations, and flag an error when an out of order cursor traversal happens.

Also add a wtperf configuration that triggers the bug reliably when run.

### CODE REVIEW

**Comment by agorrod** [comment]
The content of this branch aren't intended to be merged into develop.

**Comment by agorrod** [comment]
I can trigger an out of order cursor traversal within 10 seconds on this branch. The following is an example of the failure:

```
$ rm -rf ./WT_TEST/* && ./bench/wtperf/wtperf -O ../bench/wtperf/runners/stress-btree-split.wtperf -o verbose=2
Workload configuration:
	Home: WT_TEST
	Table name: test
	Connection configuration: create,cache_size=8GB,statistics=[fast,clear],statistics_log=(wait=10),eviction=(threads_max=4,threads_min=4),create,cache_size=8GB,statistics=[fast,clear],statistics_log=(wait=10),eviction=(threads_max=4,threads_min=4),
	Session configuration: 
	Creating new table: key_format=S,value_format=S,type=lsm,exclusive=true,allocation_size=4kb,internal_page_max=64kb,leaf_page_max=4kb,split_pct=100,type=file,leaf_page_max=8k,internal_page_max=8k,memory_page_max=2MB,split_deepen_min_child=250,key_format=S,value_format=S,type=lsm,exclusive=true,allocation_size=4kb,internal_page_max=64kb,leaf_page_max=4kb,split_pct=100,type=file,leaf_page_max=8k,internal_page_max=8k,memory_page_max=2MB,split_deepen_min_child=250,
	Key size: 20, value size: 256
	Populate threads: 2, inserting 2000000 rows
	Workload seconds, operations: 3600, 0
	Workload configuration(s):
		1 threads (inserts=1, reads=0, updates=0, truncates= 0)
		4 threads (inserts=0, reads=0, updates=1, truncates= 0)
		8 threads (inserts=0, reads=1, updates=0, truncates= 0)
	Checkpoint threads, interval: 0, 120
	Reporting interval: 5
	Sampling interval: 0
	Verbosity: 2
Starting 2 populate thread(s) for 2000000 items
Finished load of 2000000 items
Load time: 5.04
load ops/sec: 0
Starting workload #1: 1 threads, inserts=1, reads=0, updates=0, truncate=0, throttle=10000
Starting workload #2: 4 threads, inserts=0, reads=0, updates=1, truncate=0, throttle=0
Starting workload #3: 8 threads, inserts=0, reads=1, updates=0, truncate=0, throttle=0
[1450772270:441449][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR.next out-of-order keys: returning 0000000000001539230\00 which is not larger than a previously returned key 0000000000001549223\00
[1450772270:441555][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR last operations (most-recent first): next, search, next, next, next, next, next, next, next, next, next, next, next, next, next, next, next, next, next, next
[1450772270:441587][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 7 at row-next-stop: home: 0x7f3b481049a0, split-gen: 2, page: 0x7f3b1a095d10, split_action: 0, flags: 0x2
[1450772270:441617][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 6 at row-next-start: home: 0x7f3b481049a0, split-gen: 2, page: 0x7f3b1a095d10, split_action: 0, flags: 0x2
[1450772270:441639][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 5 at cursor-next: home: 0x7f3b481049a0, split-gen: 2, page: 0x7f3b1a095d10, split_action: 0, flags: 0x2
[1450772270:441658][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 4 at walk-stop: home: 0x7f3b481049a0, split-gen: 2, page: 0x7f3b1a095d10, split_action: 0, flags: 0x2
[1450772270:441676][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 3 at walk-start: home: 0x7f3b481049a0, split-gen: 2, page: 0x7f3b0c190690, split_action: 0, flags: 0x2
[1450772270:441694][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 2 at row-next-stop: home: 0x7f3b481049a0, split-gen: 2, page: 0x7f3b0c190690, split_action: 0, flags: 0x2
[1450772270:441712][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 1 at row-next-start: home: 0x7f3b481049a0, split-gen: 2, page: 0x7f3b0c190690, split_action: 0, flags: 0x2
[1450772270:441733][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 0 at cursor-next: home: 0x7f3b481049a0, split-gen: 2, page: 0x7f3b0c190690, split_action: 0, flags: 0x2
[1450772270:441752][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 19 at cursor-next-return: home: 0x7f3b481049a0, split-gen: 2, page: 0x7f3b0c190690, split_action: 0, flags: 0x2
[1450772270:441771][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 18 at row-next-stop: home: 0x7f3b481049a0, split-gen: 2, page: 0x7f3b0c190690, split_action: 0, flags: 0x2
[1450772270:441789][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 17 at row-next-start: home: 0x7f3b481049a0, split-gen: 2, page: 0x7f3b0c190690, split_action: 0, flags: 0x2
[1450772270:441806][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 16 at cursor-next: home: 0x7f3b481049a0, split-gen: 2, page: 0x7f3b0c190690, split_action: 0, flags: 0x2
[1450772270:441823][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 15 at cursor-next-return: home: 0x7f3b48078d80, split-gen: 1, page: 0x7f3b4deec5b0, split_action: 1, flags: 0x20
[1450772270:441853][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 14 at row-next-stop: home: 0x7f3b48078d80, split-gen: 1, page: 0x7f3b4deec5b0, split_action: 1, flags: 0x20
[1450772270:441872][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 13 at row-next-start: home: 0x7f3b48078d80, split-gen: 1, page: 0x7f3b4deec5b0, split_action: 1, flags: 0x20
[1450772270:441889][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 12 at cursor-next: home: 0x7f3b48078d80, split-gen: 1, page: 0x7f3b4deec5b0, split_action: 1, flags: 0x20
[1450772270:441907][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 11 at cursor-next-return: home: 0x7f3b48078d80, split-gen: 1, page: 0x7f3b4deec5b0, split_action: 1, flags: 0x20
[1450772270:441924][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 10 at row-next-stop: home: 0x7f3b48078d80, split-gen: 1, page: 0x7f3b4deec5b0, split_action: 1, flags: 0x20
[1450772270:441941][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 9 at row-next-start: home: 0x7f3b48078d80, split-gen: 1, page: 0x7f3b4deec5b0, split_action: 1, flags: 0x20
[1450772270:441964][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 8 at cursor-next: home: 0x7f3b48078d80, split-gen: 1, page: 0x7f3b4deec5b0, split_action: 1, flags: 0x20
Key[0]: 1549222
Key[1]: 1549223
Out of order keys 1549223 came before 1539230 Error: Invalid argument
[1450772270:631368][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR.next out-of-order keys: returning 0000000000000672634\00 which is not larger than a previously returned key 0000000000000682630\00
[1450772270:631406][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR last operations (most-recent first): next, next, next, next, next, next, next, next, next, next, next, next, next, next, next, next, next, next, next, next
[1450772270:631421][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 10 at row-next-stop: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b1a2d9840, split_action: 0, flags: 0x2
[1450772270:631433][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 9 at row-next-start: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b1a2d9840, split_action: 0, flags: 0x2
[1450772270:631454][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 8 at cursor-next: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b1a2d9840, split_action: 0, flags: 0x2
[1450772270:631466][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 7 at walk-stop: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b1a2d9840, split_action: 0, flags: 0x2
[1450772270:631480][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 6 at walk-start: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631493][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 5 at row-next-stop: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631511][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 4 at row-next-start: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631524][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 3 at cursor-next: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631536][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 2 at cursor-next-return: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631547][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 1 at row-next-stop: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631565][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 0 at row-next-start: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631584][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 19 at cursor-next: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631597][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 18 at cursor-next-return: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631609][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 17 at row-next-stop: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631627][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 16 at row-next-start: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631640][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 15 at cursor-next: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631661][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 14 at cursor-next-return: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631676][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 13 at row-next-stop: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631690][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 12 at row-next-start: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
[1450772270:631701][9640:0x7f3b597fa700], file:test.wt, WT_CURSOR.next: SUPPORT-1531: WT_CURSOR tracking slot 11 at cursor-next: home: 0x7f3b400807d0, split-gen: 2, page: 0x7f3b190593d0, split_action: 0, flags: 0x2
Key[0]: 682589
Key[1]: 682590
Key[2]: 682591
Key[3]: 682592
Key[4]: 682593
Key[5]: 682594
Key[6]: 682595
Key[7]: 682596
Key[8]: 682597
Key[9]: 682598
Key[10]: 682599
Key[11]: 682600
Key[12]: 682601
Key[13]: 682602
Key[14]: 682603
Key[15]: 682604
Key[16]: 682605
Key[17]: 682606
Key[18]: 682607
Key[19]: 682608
Key[20]: 682609
Key[21]: 682610
Key[22]: 682611
Key[23]: 682612
Key[24]: 682613
Key[25]: 682614
Key[26]: 682615
Key[27]: 682616
Key[28]: 682617
Key[29]: 682618
Key[30]: 682619
Key[31]: 682620
Key[32]: 682621
Key[33]: 682622
Key[34]: 682623
Key[35]: 682624
Key[36]: 682625
Key[37]: 682626
Key[38]: 682627
Key[39]: 682628
Key[40]: 682629
Key[41]: 682630
Out of order keys 682630 came before 672634 Error: Invalid argument
280356 reads, 51132 inserts, 905568 updates, 0 truncates, 0 checkpoints in 5 secs (5 total secs)
```

**Comment by sueloverso** [comment]
FYI this does not reproduce on 2.6.

**Comment by agorrod** [comment]
tl;dr I haven't got anything concrete to show for my day. I have more confidence that the scenario we talked through this morning is the cause but no red-hot smoking gun. Read on for more detail.

@keithbostic I haven't made much progress here. I've fiddled with ``__tree_walk_internal`` a bunch and have been getting more confidence that our diagnosis is right. I haven't been able to make a code change that is definitive though.

On the same path, I was hoping to figure out a set of conditions in the walk function that could be the basis for a fix, but no such luck. I have reduced the reproducer down again:

```
$ cat bench/wtperf/runners/stress-btree-split.wtperf
conn_config="cache_size=2GB,statistics=[fast,clear],statistics_log=(wait=10),eviction=(threads_max=4,threads_min=4)"
table_config="type=file,leaf_page_max=8k,internal_page_max=8k,memory_page_max=2MB,split_deepen_min_child=250"
icount=200000
report_interval=5
run_time=3600
reopen_connection=false
populate_threads=2
value_sz=256
read_range=100
threads=((count=4,inserts=1,throttle=100000),(count=8,reads=1))
```

Fires reliably for me within 2 seconds, when I've increased the sleep in ``__split_internal`` to a second.

**Comment by agorrod** [comment]
For the record, I believe what is happening is:

* A cursor traversal starts walking through a subtree. It is at a particular location in the tree.
* A number of inserts happen in parallel, triggering a split of the internal page being walked by the traversal.
* The traversal continues on to the end of the internal page (fine). When it gets to the end of the internal page, it pops up to the parent page to find the next page to traverse. The next page it retrieves was part of the split, and so part of the key range that the cursor has already walked beyond.

I've attempted to draw this in two pictures, as attached. It's kind of messy.

img_20151223_172616546

img_20151223_172931893

**Comment by keithbostic** [comment]
@agorrod, for the record, cherry-picking 8596071 appears to fix the diagnostic branch.

**Comment by keithbostic** [comment]
@agorrod, an additional race/bug in this code can be triggered with this change.
```
*** src/btree/bt_walk.c.orig      Mon Dec 14 14:46:17 2015
--- src/btree/bt_walk.c   Fri Dec 25 10:28:06 2015
***************
*** 194,199 ****
--- 194,200 ----
                if ((prev && slot == 0) ||
                    (!prev && slot == pindex->entries - 1)) {
                        ref = ref->home->pg_intl_parent_ref;
+                       __wt_sleep(1, 0);

                        /*
                         * If we got all the way through an internal page and
```
Basically, acquire the parent pointer when ascending the tree and then sleep. The failure is the same, returning out-of-order key/value pairs.

The race is:
* a thread ascends through its WT_REF to the parent,
* the WT_REFs are rewritten (that's commit 8596071),
* the parent's page index is updated,
* the thread moves through the parent's page index to the current slot,
* the thread moves to the next slot which is the result of the split.

**Comment by keithbostic** [comment]
Deleting this diagnostic pull request for now, #2415 will hopefully resolve the problem.
